### PR TITLE
Share documents

### PR DESCRIPTION
**Pull request by fid-jose**
The functionality of sharing the opened document, allowing it to be opened or sent using other apps has been added.

A content provider has been implemented so documents which might not be accessible by other means can be read by other apps.

### CODE REVIEW

File: document-viewer/src/main/java/org/ebookdroid/common/provider/DocumentContentProvider.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public boolean onCreate() {
+        return true;
+    }
+
+    @Override
+    public String getType(Uri uri) {
+        return uri.getQuery();
+    }
```
Comment (ericwa) [inline document-viewer/src/main/java/org/ebookdroid/common/provider/DocumentContentProvider.java]: According to http://developer.android.com/guide/topics/providers/content-provider-creating.html#FileMIMETypes the ContentProvider should also implement `getStreamTypes()`, if I am understanding that correctly

File: document-viewer/src/main/java/org/ebookdroid/common/provider/DocumentContentProvider.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import android.net.Uri;
+import android.os.ParcelFileDescriptor;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+
+public class DocumentContentProvider extends ContentProvider {
+
+    @Override
+    public ParcelFileDescriptor openFile(Uri uri, String mode) throws FileNotFoundException {
```
Comment (ericwa) [inline document-viewer/src/main/java/org/ebookdroid/common/provider/DocumentContentProvider.java]: I'm worried this content provider is a security vulnerability, it looks like any app could use it to read any file on the device? Going to look into it a bit more

**Comment by ericwa** [comment]
Thanks for contributing this! A share button has been on my wishlist for a while.

A few things:
- we may not want to share file:// URI's at all. Android docs suggest always using `content://` URI's, see: http://developer.android.com/training/secure-file-sharing/index.html . Also this comment in the [FileProvider](http://developer.android.com/reference/android/support/v4/content/FileProvider.html) documentation is relevant: 
   > In comparison, to control access to a file:/// Uri you have to modify the file system permissions of the underlying file. The permissions you provide become available to any app, and remain in effect until you change them. This level of access is fundamentally insecure.

   > The increased level of file access security offered by a content URI makes FileProvider a key part of Android's security infrastructure.

- I'm worried the DocumentContentProvider might allow any app to read any file. It might be better to use the `FileProvider` class from support.v4. The only complication with that is, FileProvider only allows sharing files from within a predefined directory, so we'd need to copy a document to this temporary directory before sharing it. This shouldn't be too much of a problem, and we could delete the temporary file when the ViewerActivity closes, for example.

**Comment by ericwa** [comment]
@fid-jose hope you don't mind I made a few changes and opened a new pull request here: https://github.com/SufficientlySecure/document-viewer/pull/173

I looked at using support.v4.FileProvider, but it's a pain to copy every shared document to the cache directory (support.v4.FileProvider can only share things from a predefined directory).

**Comment by fid-jose** [comment]
@ericwa Of course I don't mind :)

I'm going to close this pull request as it makes more sense to use #173

**Pull request closed without merging**